### ISSUE TITLE
make domain of objective optional

### ISSUE THREAD

**Issue by jakob-r** [mlr-org/bbotk#38]
Sometimes we only know the search_space (param_set of the OptimInstance) but not the domain (search space after trafo) of the objective. We should not always be obliged to define a domain because it is not used anyways.

Therefore it should be optional to define the domain.

**Comment by berndbischl** [comment]
This issue is really hard to understand if you don't provide a more concrete explanation.
Can you pls give a concrete example?

**Comment by jakob-r** [comment]
Look at the following example
```r
fun = function(xs) {
  y = sum(as.numeric(xs)^2)
  list(y = y)
}

ps_domain = ParamSet$new(list(
  ParamDbl$new("x1"),
  ParamDbl$new("x2")
))

ObjectiveRFun$new(fun = fun, domain = ps_domain)

ps_search_space = ParamSet$new(list(
  ParamDbl$new("x1", lower = -1, upper = 1),
  ParamDbl$new("x2", lower = -1, upper = 1)
))

tt = TerminatorEvals$new()
OptimInstance$new(
  objective = objective, 
  search_space = ps_search_space, 
  terminator = tt)
```

I could use `ps_domain <- ps_search_space` but this looks unnecessary complicated. So why do we have it twice? Because `ps_domain` can be a superset. However, in the code we never use `ps_domain` to check anything. Yes we have `eval_checked` but it is never called from the `OptimInstance` and nobody is going to use the objective directly anyhow. So I just want to strip down the complexity here.

**Comment by jakob-r** [comment]
Or we make the `search_space` optional for the `OptimInstance` and use the `domain` as the default.

**Comment by berndbischl** [comment]
> Yes we have `eval_checked` but it is never called from the `OptimInstance`

just saying: that's kinda bad right? we should use that?

**Comment by berndbischl** [comment]
I dunno. Removing the domain doesnt really reduce complexity, it makes it MORE complex? 
You now cannot rely on it being there. See "eval_checked". You would have to remove those functions now, or somehow say they cannot be called? 
The domain of a function should really always be known by the user? We can maybe talk about how to specify them faster?

**Comment by berndbischl** [comment]
> Or we make the `search_space` optional for the `OptimInstance` and use the `domain` as the default.

that OTOH I could agree with?

**Comment by berndbischl** [comment]
we are doint the last, @be-marc also likes it, that seems clean and easy

**Comment by be-marc** [comment]
Closed by #91
